Implement advanced verification of SD-JWT in Keycloak

IngridPuppet opened the pull request:
Closes #30907

IngridPuppet committed:
Implement advanced verification of SD-JWT in Keycloak

Closes #30907

Signed-off-by: Ingrid Kamga <Ingrid.Kamga@adorsys.com>

mposolda commented:
@wistefan @francis-pouatcha @tnorimat Anyone of you available to review this please?

tnorimat commented:
@mposolda Some test failed, but it seems that these failed tests are not relevant to the PR. Could you re-run the CI for the PR?

tnorimat commented:
@mposolda The re-ran CI also had one fail, but the fail is different from the one caused by the previous-ran CI.
Maybe, the fail was not caused by the PR.

ahus1 commented:
The CI rerun for the job failed as it was missing the archive which was expired. I rerun the full CI. Let's see if that succeeds.

ahus1 commented:
@wistefan @francis-pouatcha @tnorimat - the tests are now green. Please continue your review when you have the time.

tnorimat commented:
@ahus1 Thank you. I will review the PR.

tnorimat commented on core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java:
```diff
@@ -0,0 +1,424 @@
+                .withValidateNotBeforeClaim(false);
+    }
+
+    private ObjectNode exampleS20KbPayload() {
```
What does `S20` means? If possible, could you name the method that other people can understand?

IngridPuppet replied:
I've removed the `S20` particle to make it less obscure. The method has no further intent beyond pre-constructing a key binding jwt payload.

tnorimat reviewed (changes requested):
@IngridPuppet Thank you for the PR. I added some review comments. Could you check them?

IngridPuppet commented on core/src/main/java/org/keycloak/sdjwt/SdJws.java:
```diff
@@ -107,4 +107,65 @@ private static final JsonNode readPayload(JWSInput jwsInput) {
+    }
+
+    /**
+     * Verifies that SD-JWT vct claim matches the expected one. Verification is case-insensitive
```
I've removed that relaxation. It is now up to the caller to possibly match case-insensitively, according to its policy.

IngridPuppet commented on core/src/test/java/org/keycloak/sdjwt/SdJwtVerificationTest.java:
```diff
@@ -0,0 +1,418 @@
+                .withSigner(testSettings.issuerSigContext);
+    }
+
+    private SdJwt.Builder exampleSdJwtWithUndisclosedNestedFieldsV1() {
```
I acknowledge that the methods are in fact largely similar, but they are not the same. Method `exampleRecursiveSdJwtV1()` has this line:

```
.withUndisclosedClaim("address", "BZFzhQsdPfZY1WSL-1GXKg")
```

By making the whole address object selectively disclosable, we can test that the implemented logic can recursively process the object accordingly.

To address this point, I've refactored the methods to decrease code duplication and added a comment above the key line addition.

tnorimat commented on core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:
```diff
@@ -0,0 +1,656 @@
+
+        // Build verifier
+        if (keyWrapper.getType().equals(KeyType.EC)) {
+            if (keyWrapper.getAlgorithm() == null) {
```
Sorry, it is my misunderstanding.

IngridPuppet commented on core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:
```diff
@@ -0,0 +1,656 @@
+     *
+     * @return the fully disclosed SdJwt payload
+     */
+    private JsonNode validateViaRecursiveDisclosing(
```
I agree that enforcing this rule should also be implemented. I have added the code for it.

IngridPuppet commented on core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:
```diff
@@ -0,0 +1,656 @@
+                keyWrapper.setAlgorithm(alg);
+            }
+
+            return new ECDSASignatureVerifierContext(keyWrapper);
```
Sorry, I made a mistake while searching for the available implementations of the `SignatureVerifierContext` interface in the codebase. I've now added support for RSA key binding signatures.

As for EdDSA, while there is a corresponding `SignatureVerifierContext` implementation, I am afraid there isn't yet support for parsing OKP JWKs into a `KeyWrapper`, a feature on which the current logic relies.

https://github.com/keycloak/keycloak/blob/073a6373628489b65e0d192e075957d58ed371ea/core/src/main/java/org/keycloak/jose/jwk/AbstractJWKParser.java#L114-L116

I additionally think that this PR is not the right place to bridge this gap. I would however welcome any further guidance.

IngridPuppet commented on core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java:
```diff
@@ -0,0 +1,424 @@
+        }
+    }
+
+    private void testShouldFailGenericS20(
```
I've renamed the method to `testShouldFailGeneric2` to make it less obscure, then added a comment to specify that the base sdjwt it considers is sample `sdjwt/s20.1-sdjwt+kb.txt`.

IngridPuppet commented on core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java:
```diff
@@ -0,0 +1,424 @@
+
+    @Test
+    public void testVerif_s20_1_sdjwt_with_kb() throws VerificationException {
+        String sdJwtVPString = TestUtils.readFileAsString(getClass(), "sdjwt/s20.1-sdjwt+kb.txt");
```
This point and the ones alike below are concerned about the origin of the following files used in the testsuite:

- s20.1-sdjwt+kb.txt
- s20.1-sdjwt+kb--wrong-kb-signature.txt
- s20.2-sdjwt+kb--no-cnf-claim.txt
- s20.3-sdjwt+kb--wrong-kb-typ.txt
- s20.6-sdjwt+kb--disclosure-with-no-digest.txt
- s20.7-sdjwt+kb--invalid-array-elt-disclosure.txt
- s20.7-sdjwt+kb--invalid-field-disclosure.txt
- s20.8-sdjwt+kb--cnf-hmac.txt
- s20.8-sdjwt+kb--cnf-is-not-jwk.txt
- s20.8-sdjwt+kb--cnf-jwk-is-malformed.txt
- s20.8-sdjwt+kb--cnf-rsa-ps256.txt
- s20.8-sdjwt+kb--cnf-rsa-ps512.txt
- s20.8-sdjwt+kb--cnf-rsa-rs256.txt
- s20.8-sdjwt+kb--es384.txt
- s20.8-sdjwt+kb--es512.txt

All of them were manually crafted. The first one pairs an sdjwt that was present in the project with a corresponding key binding JWT issued in another test. The other ones use online tools such as jwt.io to create alterations for alternative scenarios.

I have added a global comment to specify that. Is that appropriate?

IngridPuppet committed:
Address requested changes

Signed-off-by: Ingrid Kamga <Ingrid.Kamga@adorsys.com>

tnorimat reviewed (commented):
@IngridPuppet Thank you. Could you check review comments?

IngridPuppet commented on core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:
```diff
@@ -0,0 +1,709 @@
+    ) throws VerificationException {
+        JsonNode nonce = keyBindingJwt.getPayload().get("nonce");
+        if (nonce == null || !nonce.isTextual()
+                || !nonce.asText().equals(keyBindingJwtVerificationOpts.getNonce())) {
```
Do you mean that the wallet reads the `nonce` from the authorization request when creating the key binding JWT? Yes. Here is how we instantiate `keyBindingJwtVerificationOpts` in a project:

```java
var nonce = requestObjectDTO.getNonce();
var clientId = requestObjectDTO.getClientId();
        
var keyBindingJwtVerificationOpts = keyBindingJwtVerificationOptsBuilder
        .withKeyBindingRequired(isKeyBindingRequired)
        .withAud(clientId)
        .withNonce(nonce)
        .build();
```

tnorimat replied:
@IngridPuppet Thank you for the clarification.

IngridPuppet commented on core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:
```diff
@@ -0,0 +1,709 @@
+
+        JsonNode aud = keyBindingJwt.getPayload().get("aud");
+        if (aud == null || !aud.isTextual()
+                || !aud.asText().equals(keyBindingJwtVerificationOpts.getAud())) {
```
Do you mean that the wallet read the `aud` value from the authorization request (field `client_id`) when creating the key binding JWT? Yes. Here is how we instantiate `keyBindingJwtVerificationOpts` in a project:

```java
var nonce = requestObjectDTO.getNonce();
var clientId = requestObjectDTO.getClientId();
        
var keyBindingJwtVerificationOpts = keyBindingJwtVerificationOptsBuilder
        .withKeyBindingRequired(isKeyBindingRequired)
        .withAud(clientId)
        .withNonce(nonce)
        .build();
```

tnorimat replied:
@IngridPuppet Thank you for the clarification.

IngridPuppet commented on core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:
```diff
@@ -0,0 +1,709 @@
+     */
+    private void validateDisclosuresVisits(Set<String> visitedDisclosureStrings)
+            throws VerificationException {
+        if (visitedDisclosureStrings.size() < disclosures.size()) {
```
Yes, you're right: we want to check that all disclosures have been visited. So when the condition is NOT met, then we throw an exception.

IngridPuppet committed:
Address requested changes

Signed-off-by: Ingrid Kamga <Ingrid.Kamga@adorsys.com>

IngridPuppet commented on core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:
```diff
@@ -0,0 +1,709 @@
+            throw new VerificationException("Key binding JWT: Invalid `iat` claim", e);
+        }
+
+        long now = Instant.now().getEpochSecond();
```
Agree. To further promote re-usability, I've created an appropriate method in the `SdJws` superclass.

IngridPuppet commented:
@tnorimat Upon re-running tests locally, I realized there was an issue with the `WildFlyElytronProvider` not matching RSASSA-PSS signatures with JavaAlgorithm's labels. In your review, please pay extra attention to the updates I made to make it work for SD-JWT verification.

tnorimat reviewed (approved):
LGTM

tnorimat commented:
@mposolda @wistefan @francis-pouatcha If possible, could you also check the PR?

mposolda commented:
Looks good to me, but has limited knowledge of oid4vc code

@francis-pouatcha Can you perhaps review as well please?

francis-pouatcha commented:
@mposolda  I will review this PR tomorrow.

francis-pouatcha commented:
Looked all good for me. Major feedback already addressed by @tnorimat. Requested merged of current main branch.

IngridPuppet committed:
Ensure verification runs immutably

Signed-off-by: Ingrid Kamga <Ingrid.Kamga@adorsys.com>

IngridPuppet committed:
Address remnant merge conflict

Signed-off-by: Ingrid Kamga <Ingrid.Kamga@adorsys.com>

IngridPuppet commented:
Thank you for your feedback. Some updates:
- Rebased on main and fixed the change that was breaking the compilation, at least locally — @mposolda.
- Noticed that verification couldn't be repeated because state was changed. To address this, I've deep cloned a problematic argument in commit https://github.com/keycloak/keycloak/pull/30966/commits/2e4f38c792aaa37c10d4f00d410b1d889bcf1b60. Please check again — @francis-pouatcha @tnorimat.

francis-pouatcha commented:
@IngridPuppet [2e4f38c](https://github.com/keycloak/keycloak/commit/2e4f38c792aaa37c10d4f00d410b1d889bcf1b60) looks good for me.

tnorimat commented on core/src/main/java/org/keycloak/sdjwt/SdJwtVerificationContext.java:
```diff
@@ -0,0 +1,730 @@
+                Objects.requireNonNull(keyWrapper.getCurve());
+
+                String alg = null;
+                switch (keyWrapper.getCurve()) {
```
This switch clause can be represented simply as follows.
```
                String alg = switch (keyWrapper.getCurve()) {
                    case "P-256" -> "ES256";
                    case "P-384" -> "ES384";
                    case "P-521" -> "ES512";
                    default -> null;
                };
```

IngridPuppet replied:
Thank you for recommending this simpler syntax. Actually, my IDE has also been prompting for this. However, as the `core` module in locked in Java 11, we cannot use switch expressions.
https://github.com/keycloak/keycloak/blob/adb2af442a88b88452cd8aee2d064c906d4eb9ec/core/pom.xml#L34-L39

tnorimat replied:
Thank you for the clarification.

tnorimat reviewed (approved):
@IngridPuppet Thank you. I approved the PR. I added one review comment but it is up to you to accept my recommendation for coding in the comment.

tnorimat commented:
@mposolda Hello, I and @francis-pouatcha  approved the PR. Could you check the PR?

mposolda reviewed (approved):
@IngridPuppet @tnorimat @francis-pouatcha Thanks for all the updates and for the review of this PR!

I hope to merge if tests are OK

mabartos commented:
@mposolda just one thing - might be good to mention it in the release notes, right?
